controlplane/kubeadm/controllers/remediation.go

```diff
@@ -0,0 +1,180 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controlplane/kubeadm/internal"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// reconcileUnhealthyMachines handles KubeadmControlPlane unhealthyMachine reconciliation.
+func (r *KubeadmControlPlaneReconciler) reconcileUnhealthyMachines(ctx context.Context, c *internal.ControlPlane) (ret ctrl.Result, retErr error) {
+	logger := r.Log.WithValues("namespace", c.KCP.Namespace, "kubeadmControlPlane", c.KCP.Name, "cluster", c.Cluster.Name)
+
+	// Gets all machines that have a MachineHealthCheckSucceeded condition set to False,
+	// indicating a problem was detected on the machine, and the MachineOwnerRemediated condition set, indicating that KCP is
+	// responsible of performing remediation as owner of the machine.
+	unhealthyMachines := c.UnhealthyMachines()
+
+	// If there are not unhealthy machines, return so KCP can proceed with other operations (ctrl.Result nil).
+	if len(unhealthyMachines) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// Select the machine to be remediated, which is the oldest machine marked as unhealthy.
+	// NOTE: The current solution is considered acceptable for the most frequent use case (only one unhealthy machine),
+	// however, in the future this could potentially be improved for the scenario where more than one unhealthy machine exists
+	// by considering which machine has lower impact on etcd quorum.
+	machineToBeRemediated := unhealthyMachines.Oldest()
+
+	// Returns if the machine is already being remediated
+	if !machineToBeRemediated.ObjectMeta.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	patchHelper, err := patch.NewHelper(machineToBeRemediated, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to Patch the Machine conditions after each reconcileUnhealthyMachines.
+		if err := patchHelper.Patch(ctx, machineToBeRemediated, patch.WithOwnedConditions{Conditions: []clusterv1.ConditionType{
+			clusterv1.MachineOwnerRemediatedCondition,
+		}}); err != nil {
+			logger.Error(err, "Failed to patch control plane Machine", "machine", machineToBeRemediated.Name)
+			if retErr == nil {
+				retErr = errors.Wrapf(err, "failed to patch control plane Machine %s", machineToBeRemediated.Name)
+			}
+		}
+	}()
+
+	// before starting remediation, run preflight checks in order to verify it is safe to remediate.
+	// If those checks are failing, we surface the reason why remediation is not happening into the MachineOwnerRemediated condition
+	// and then we return so KCP can proceed with other operations (ctrl.Result nil).
+
+	var desiredReplicas int = 0
```

Comment:
Given that `Spec.Replicas` defaults to 1.
```suggestion
	var desiredReplicas int = 1
```

Reply:
Do we even need this check? If there is defaulting in place, we should be able to assume that we can just do 
```suggestion
	desiredReplicas := *c.KCP.Spec.Replicas
```